fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownAggregateCollector.java

```diff
@@ -0,0 +1,550 @@
+                    if (caseWhen.getThenClause(i).isConstant() && !caseWhen.getThenClause(i).isConstantNull()) {
+                        // forbidden push down
+                        return visit(optExpression, context);
+                    }
+                    newWhenThen.add(ConstantOperator.createBoolean(false));
+                    newWhenThen.add(caseWhen.getThenClause(i));
+                }
+
+                // mock just value case when
+                CaseWhenOperator newCaseWhen = new CaseWhenOperator(caseWhen.getType(), null,
+                        caseWhen.hasElse() ? caseWhen.getElseClause() : null, newWhenThen);
+
+                // replace origin
+                aggFn.setChild(0, newCaseWhen);
+            } else if (callInput.getFunction() != null &&
+                    FunctionSet.IF.equals(callInput.getFunction().getFunctionName().getFunction())) {
+                if (aggInput.getChildren().stream().skip(1).anyMatch(c -> c.isConstant() && !c.isConstantNull())) {
+                    // forbidden push down
+                    return visit(optExpression, context);
+                }
+
+                aggInput.getChild(0).getUsedColumns().getStream().mapToObj(factory::getColumnRef)
+                        .forEach(v -> context.groupBys.put(v, v));
+                aggInput.setChild(0, ConstantOperator.createBoolean(false));
+            }
+        }
+
+        // check has constant aggregate, forbidden
+        if (context.aggregations.values().stream().anyMatch(ScalarOperator::isConstant)) {
+            return visit(optExpression, context);
+        }
+
+        return processChild(optExpression, context);
+    }
+
+    @Override
+    public Void visitLogicalAggregate(OptExpression optExpression, AggregateContext context) {
+        LogicalAggregationOperator aggregate = (LogicalAggregationOperator) optExpression.getOp();
+        // distinct/count*/constant aggregate can't push down
+        if (aggregate.getAggregations().values().stream()
+                .anyMatch(c -> c.isDistinct() || c.isCountStar() || c.isConstant())) {
+            return visit(optExpression, context);
+        }
+
+        context = new AggregateContext();
+        context.setAggregator(aggregate);
+        return processChild(optExpression, context);
+    }
+
+    @Override
+    public Void visitLogicalJoin(OptExpression optExpression, AggregateContext context) {
+        if (isInvalid(optExpression, context)) {
+            return visit(optExpression, context);
+        }
+        LogicalJoinOperator join = (LogicalJoinOperator) optExpression.getOp();
+        // constant aggregate can't push down
+        if (context.aggregations.values().stream().anyMatch(ScalarOperator::isConstant)) {
+            return visit(optExpression, context);
+        }
+
+        if (join.getOnPredicate() != null) {
+            join.getOnPredicate().getUsedColumns().getStream().mapToObj(factory::getColumnRef)
```

Comment:
columnRef of non-equal predicate need to be added into agg's group-by columns?

Reply:
need, should add columns no matter what predicate

Reply:
Could you add some comments to explain the conditions of agg pushing down to join